### ISSUE TITLE
sess.run([train_step]) freezes when using batch_normalization and Collection Update

### ISSUE THREAD

**Issue by amirmazaheri1990** [tensorflow/tensorflow#19568]
Hi,

I have faced a strange situation with Tensorflow. I explored a lot about this problem but just found one other thread (unsolved) on StackOverflow. (here: https://stackoverflow.com/questions/47047124/tf-layers-batch-normalization-freezes-during-sess-run-1-5-0-dev20171031). So, I decided to ask it here.

Basically, when I call sess.run(), it freezes. By freeze, I mean, the GPU utilization is zero, I get no errors and the process is on GPU (GPU memory is allocated).  I have the following code segment:
```
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
  with tf.control_dependencies(update_ops):
        train_step1 = optimizer1.minimize(loss = loss_fill +lossL2+ loss_detection,var_list=vars)
```

When I change this part to:
```
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
     with tf.control_dependencies(update_ops):
             pass
train_step1 = optimizer1.minimize(loss = loss_fill +lossL2+ loss_detection,var_list=vars)
```

which basically ignores some necessary moving average updates, it doesn't freeze anymore. I have a lot of tf.layers.batch_normalization() instances in my code and this is the first time I am facing this issue.

Thanks

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by facaiy** [comment]
It looks strange that you put `update_op` before `minimize`.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sharvil** [comment]
I'm seeing this issue as well. Here's my configuration:

Ubuntu 16.04.5
TF 1.9.0 (v1.9.0-0-g25c197e) from source
Bazel 0.15.2
CUDA Version 9.1.85
cuDNN 7.1.3
GeForce GTX 1080 Ti 11GB

The following sample code demonstrates the issue:

```python
import tensorflow as tf
import numpy as np

BATCH_SIZE = 1
TIME_STEPS = 32  # no lock up with TIME_STEPS < 32

memory = tf.placeholder(tf.float32, [BATCH_SIZE, 1, 3])

cell = tf.contrib.seq2seq.AttentionWrapper(
  tf.contrib.rnn.LSTMCell(3),
  tf.contrib.seq2seq.BahdanauAttention(3, memory),
)

inputs = np.zeros((BATCH_SIZE, TIME_STEPS, 1), dtype=np.float32)
helper = tf.contrib.seq2seq.TrainingHelper(inputs, [TIME_STEPS])
initial_state = cell.zero_state(BATCH_SIZE, tf.float32)
decoder = tf.contrib.seq2seq.BasicDecoder(cell, helper, initial_state)

final_output, _, _ = tf.contrib.seq2seq.dynamic_decode(decoder)

res = final_output.rnn_output
res = tf.layers.batch_normalization(res, training=True)

global_step = tf.train.get_or_create_global_step()
optimizer = tf.train.AdamOptimizer()

with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
  grads_and_vars = optimizer.compute_gradients(res)
  train_op = optimizer.apply_gradients(grads_and_vars, global_step)

sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(train_op, feed_dict={memory: [[[1, 2, 3]]]})
```

There seems to be a connection between the number of decoder steps and batch normalization. The issue is reproducible with `TIME_STEPS >= 32` but not with `TIME_STEPS < 32`.

@mrry, could you please help find someone to triage this issue? Thanks!

**Comment by sharvil** [comment]
Friendly ping. Checking in to see if we can get this bug triaged.

**Comment by sharvil** [comment]
@raofengyun thanks for the suggestion! You're right that eliminating the `control_dependency` will prevent the lockup from occurring. However, `update_ops` isn't guaranteed to complete before `train_op` even if it's placed first in the `session.run` call unless that control dependency is introduced.

I've also noticed that the above code snippet runs to completion when executing on CPU.

**Comment by tfboyd** [comment]
@tfboyd

**Comment by raofengyun** [comment]
@sharvil you're right, it's not guaranteed, but I can't find any other way.

**Comment by sharvil** [comment]
@asimshankar, @tfboyd any update on this issue?

**Comment by asimshankar** [comment]
@sharvil : Apologies for the late response. Nope, no update yet, haven't gotten around to debugging this yet. 

@zheng-xq @chsigg - would either of you have some cycles to look into this?
@drpngx @qlzh727 have been looking at the LSTMCell implementations recently and may also have some insight.

For the record, if I understand correctly, the following reproduces the problem (picked up from the [StackOverflow question](https://stackoverflow.com/questions/47047124/tf-layers-batch-normalization-freezes-during-sess-run-1-5-0-dev20171031), haven't tried it myself yet):

```python
import tensorflow as tf
import numpy as np

starter_learning_rate = 0.001
decay_steps = 100
decay_rate = 0.96
num_RNN_layers = 3
LSTM_CELL_SIZE = 100
keep_prob = 0.95

with tf.name_scope('Inputs'):
    x = tf.placeholder(dtype=tf.float32, shape=[None, 200])
    y = tf.placeholder(dtype=tf.float32, shape=[None, 200])
    length = tf.placeholder(dtype=tf.int32, shape=[None])
    Flg_training = tf.placeholder(dtype=tf.bool, shape=[])

    x_1 = tf.expand_dims(x, -1)

with tf.name_scope('BiLSTM'):
    dropcells = []
    for iiLyr in list(range(num_RNN_layers)):
        cell_iiLyr = tf.nn.rnn_cell.LSTMCell(num_units=LSTM_CELL_SIZE, state_is_tuple=True)
        dropcells.append(tf.nn.rnn_cell.DropoutWrapper(cell=cell_iiLyr, output_keep_prob=keep_prob))  #,, input_keep_prob=self.keep_prob input_keep_prob=1.0, seed=None

    MultiLyr_cell = tf.nn.rnn_cell.MultiRNNCell(cells=dropcells, state_is_tuple=True)

    outputs, states  = tf.nn.bidirectional_dynamic_rnn(
        cell_fw=MultiLyr_cell, 
        cell_bw=MultiLyr_cell, 
        dtype=tf.float32,
        sequence_length=length, #tf_b_lens 
        inputs=x_1, #stacked_RefPts_desc, #tf_b_VCCs_AMs_BN1
        scope = "BiLSTM"
        )

    #output_fw, output_bw = outputs
    states_fw, states_bw = states

    c_fw_lstLyr, h_fw_lstLyr = states_fw[-1]
    c_bw_lstLyr, h_bw_lstLyr = states_bw[-1]

    states_concat1 = tf.concat([h_fw_lstLyr, h_bw_lstLyr], axis = 1, name = 'states_concat')

with tf.name_scope("cs_BN1"):
    x_BN = tf.layers.batch_normalization(
        states_concat1,
        axis=-1, # axis that should be normalized (typically the features axis, in this case the concated states or hidden vectors)
        momentum=0.99,
        epsilon=1e-10, #0.001,
        center=False, #True,
        scale=False, #True,
        beta_initializer=tf.zeros_initializer(),
        gamma_initializer=tf.ones_initializer(),
        moving_mean_initializer=tf.zeros_initializer(),
        moving_variance_initializer=tf.ones_initializer(),
        beta_regularizer=None,
        gamma_regularizer=None,
        beta_constraint=None,
        gamma_constraint=None,
        training=Flg_training, #False,
        trainable=True,
        name="test_BN", #None,
        reuse=None,
        renorm=False,
        renorm_clipping=None,
        renorm_momentum=0.99,
        fused=False,
        virtual_batch_size=None,
        adjustment=None
        )

with tf.name_scope("Regression"):
    a = tf.get_variable("a", shape=[1], dtype=tf.float32, initializer=tf.constant_initializer(1.0))
    b = tf.get_variable("b", shape=[1], dtype=tf.float32, initializer=tf.constant_initializer(0.0))

with tf.name_scope("Prediction"):
    y_pred = tf.multiply(x_BN, a) + b

with tf.name_scope('Loss'):
    losses = tf.losses.mean_squared_error(y, y_pred, reduction=tf.losses.Reduction.NONE)
    mean_loss = tf.reduce_mean(losses)

with tf.name_scope('Training'):
    global_step = tf.Variable(0, trainable=False)
    learning_rate = tf.train.exponential_decay(starter_learning_rate, global_step,
                                               decay_steps, decay_rate, staircase=True) 

    update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)

    with tf.control_dependencies(update_ops):
        train_step = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(losses, global_step=global_step)


#x_mean = tf.reduce_mean(x_BN, axis=0)

sess = tf.InteractiveSession()
train_writer = tf.summary.FileWriter("G:\\Surface_Ozone\\Temp\\", sess.graph)   
sess.run(tf.global_variables_initializer())

for ii in list(range(2000)):
    x_in = (np.random.rand(20, 200))
    y_in = x_in * 1.5 + 3.0
    length_in = np.full([20], 200, dtype=np.int32)

    _, mean_loss_val, a_val, b_val = sess.run([train_step, mean_loss, a, b], feed_dict={
        x: x_in, 
        Flg_training: True, 
        y: y_in,
        length: length_in
        })

    if (ii < 50):
        print("step {}: {} | a: {} | b: {}".format(ii, mean_loss_val, a_val, b_val))
    else:
        if (ii % 100 == 0):
            print("step {}: {} | a: {} | b: {}".format(ii, mean_loss_val, a_val, b_val))

print("Normal End.")
```

**Comment by sharvil** [comment]
Thanks for the update, @asimshankar. I posted a shorter code snippet that reproduces the problem in one of the comments above. You can copy/paste that in an interactive python session and verify in a few seconds.

**Comment by qlzh727** [comment]
Thanks @sharvil. I quickly try the snippet and was able to reproduce the issue with tf-nightly-gpu binary. Interestingly, the issue is only produced on GPU version, but not tf CPU binary, which I suspect has something to do with the GPU ops. 

Will need some time to dig more for details.

**Comment by sharvil** [comment]
@qlzh727 thanks for taking a look. I observed the GPU/CPU behavior difference in a comment above. Allow me to summarize what's already in this thread:

1) this issue has been present since at least TF 1.5 and has been repro'd now on 1.8, 1.9, and tf-nightly-gpu.
2) this issue only occurs when executing on GPU; using a CPU build or running with no GPU devices succeeds.
3) this issue only occurs when `TIME_STEPS` is >= 32.

My suspicion is that there's a synchronization issue in the GPU code path. Consider different values of `parallel_iterations` on `dynamic_decode`.

**Comment by sharvil** [comment]
Checking in to see if there's any update.

**Comment by sharvil** [comment]
@ppwwyyxx you're right, that issue looks suspiciously similar and probably has the same root cause. On another note, I've come up with a workaround for the sample code I posted above.

The following code:

```python
with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
  grads_and_vars = optimizer.compute_gradients(res)
  train_op = optimizer.apply_gradients(grads_and_vars, global_step)
```

can be transformed into:

```python
res = tf.tuple([res], control_inputs=tf.get_collection(tf.GraphKeys.UPDATE_OPS))[0]
grads_and_vars = optimizer.compute_gradients(res)
train_op = optimizer.apply_gradients(grads_and_vars, global_step)
```

which forces the batch norm ops to run before gradients on `res` are computed, as desired.

**Comment by jbed94** [comment]
I had similar issue with this magical number 32. Length of the sequence which i passed into rnn was always 70 but bidirectional rnn worked only when i set manually max length to 32. In other case tf hangs. I changed `parallel_iterations` argument to 71 (max size + 1) and it is working.

Unfortunately i have no idea if number of parallel iterations affects somehow the results and how does the performance changes looks like but if I see such problems, expect the worst.

Previous code:
`outputs, _ = tf.nn.bidirectional_dynamic_rnn(lstm_fw_cell, lstm_bw_cell, rnn_input, dtype=tf.float32)`

New code:
`outputs, _ = tf.nn.bidirectional_dynamic_rnn(lstm_fw_cell, lstm_bw_cell, rnn_input, dtype=tf.float32, parallel_iterations=71)`

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@amirmazaheri1990,
The [code ](https://github.com/tensorflow/tensorflow/issues/19568#issuecomment-409361140) mentioned by @sharvil  could be executed successfully in the **`Tensorflow Version 1.15.2`**, in **`Google Colab`** with **`GPU Runtime`**. Please find [the Gist](https://colab.research.google.com/gist/rmothukuru/6ce80d423941eec3d5d688c704f7a269/gh_19568.ipynb) of the working code. Thanks!

**Comment by sharvil** [comment]
Cool cool. I'm relieved this got sorted out.
